File: contents/tree_traversal/code/javascript/tree.js
function createTree(rows, children) {
  if (rows === 0) {
    return { id: rows, children: [] };
  }

  return {
    id: rows,
    children: [...Array(children).keys()].map(() => createTree(rows - 1, children))
  };
}

function dfsPreorder(tree) {
  if (!tree) {
    return;
  }

  process.stdout.write(tree.id + " ");
  tree.children.forEach(dfsPreorder);
}

function dfsPostorder(tree) {
  if (!tree) {
    return;
  }

  tree.children.forEach(dfsPostorder);
  process.stdout.write(tree.id + " ");
}

function dfsInorder(tree) {
  if (!tree) {
    return;
  }

  switch (tree.children.length) {
    case 2:
      dfsInorder(tree.children[0]);
      console.log(tree.id);
      dfsInorder(tree.children[1]);
      break;
    case 1:
      dfsInorder(tree.children[0]);
      console.log(tree.id);
      break;
    case 0:
      console.log(tree.id);
      break;
    default:
      throw new Error("Postorder traversal is only valid for binary trees");
  }
}

function dfsIterative(tree) {
  const stack = [tree];
  while (stack.length > 0) {
    const current = stack.pop();
    process.stdout.write(current.id + " ");
    stack.push(...current.children);
  }
}

function bfs(tree) {
  const queue = [tree];
  while (queue.length > 0) {
    const current = queue.shift();
    process.stdout.write(current.id + " ");
    queue.push(...current.children);
  }
}

const root = createTree(2, 3);
console.log("[#]\nRecursive DFS:");
dfsPreorder(root);
console.log();
console.log("[#]\nRecursive Postorder DFS:");
dfsPostorder(root);
console.log();
console.log("[#]\nStack-based DFS:");
dfsIterative(root);
console.log();
console.log("[#]\nQueue-based BFS:");
bfs(root);
console.log();
const root_binary = createTree(3, 2);
console.log("[#]\nRecursive Inorder DFS for Binary Tree:");
dfsInorder(root_binary);
console.log();

File: contents/tree_traversal/code/ruby/tree.rb
def create_tree(rows, children)
  return { id: rows, children: [] } if rows.zero?

  {
    id: rows,
    children: Array.new(3, create_tree(rows - 1, children))
  }
end

def dfs_preorder(tree)
  puts tree[:id]
  tree[:children].each { |child| dfs_preorder(child) }
end

def dfs_postorder(tree)
  tree[:children].each { |child| dfs_postorder(child) }
  puts tree[:id]
end

def dfs_inorder(tree)
  return unless tree

  if tree[:children].count > 2
    raise 'Postorder traversal is only valid for binary trees'
  end

  dfs_inorder(tree.children[0])
  puts tree[:id]
  dfs_inorder(tree.children[1])
end

def dfs_iterative(tree)
  stack = [tree]
  while stack.count.positive?
    current = stack.pop
    puts current[:id]
    stack.push(*current[:children])
  end
end

def bfs(tree)
  queue = [tree]
  while queue.count.positive?
    current = queue.shift
    puts current[:id]
    queue.push(*current[:children])
  end
end

root = create_tree(3, 3)
dfs_preorder(root)
dfs_postorder(root)
dfs_iterative(root)
bfs(root)
